lib/utilities/app.state.dart

```diff
@@ -25,170 +29,200 @@ class AppStateHandler extends ChangeNotifier {
+      );
       rechteProgressNotifier.value = await getRechte();
+      syncState = SyncState.successful;
+      setReadyState();
+    } on SessionExpired catch (_) {
+      debugPrint('sync failed with session expired');
+      syncState = SyncState.relogin;
+      if (await setReloginState()) {
+        if (!background) {
+          /// pop with false to prevent going to ready or loggedOut state
+          navigatorKey.currentState!.pop();
+        }
+        setLoadDataState(loadAll: loadAll, background: background);
+      } else {
+        syncState = SyncState.error;
+      }
     } catch (e) {
-      statusGreenNotifier.value = false;
-      statusGreenNotifier.notifyListeners();
-      debugPrint(e.toString());
+      if (e is http.ClientException) {
+        debugPrint('sync failed with no internet connection');
+        syncState = SyncState.offline;
+        setReadyState();
+      } else {
+        debugPrint('sync failed with error:$e');
+        syncState = SyncState.error;
+      }
     }
   }
 
   /// App is authenticated | User loggin unclear
-  /// Hive is open
-  Future<void> setAuthenticatedState(BuildContext context) async {
+  Future<void> setAuthenticatedState() async {
     if (currentState == AppState.authenticated) return;
 
-    await registerAdapter();
-    await openHive();
+    // TODO: Ask for app authentication here
 
     currentState = AppState.authenticated;
 
-    // check if data is available -> Logout if not
-    List<Mitglied> mitglieder =
-        Hive.box<Mitglied>('members').values.toList().cast<Mitglied>();
-
-    final thirtyDaysAgo = DateTime.now().subtract(const Duration(days: 30));
-
-    if (mitglieder.isEmpty) {
-      WidgetsBinding.instance.addPostFrameCallback((_) {
-        setLoggedOutState(context);
-      });
-      return;
-    }
-
-    // login check or data is older than 30 days -> load data
-    if (getLastLoginCheck().isBefore(thirtyDaysAgo) ||
-        getLastNamiSync().isBefore(thirtyDaysAgo)) {
-      WidgetsBinding.instance.addPostFrameCallback((_) {
-        setLoadDataState(context);
-      });
-      return;
-    }
-
-    WidgetsBinding.instance.addPostFrameCallback((_) {
-      setReadyState(context);
-    });
+    setReadyState();
   }
 
   /// App is authenticated | User is logged in
-  /// Hive is open
   /// Data is available and up to date
-  void setReadyState(BuildContext context) {
-    if (currentState == AppState.ready) return;
-
+  void setReadyState() {
+    final diffToLastSync = getLastNamiSync()
+        .add(const Duration(days: 1))
+        .difference(DateTime.now());
+    if (diffToLastSync < const Duration()) {
+      debugPrint("Sync data in background now");
+      setLoadDataState(background: true);
+    } else {
+      debugPrint("Scheduled sync data in $diffToLastSync in background");
+      syncTimer?.cancel();
+      syncTimer = Timer(diffToLastSync, () {
+        debugPrint("Sync data from timer now");
+        setLoadDataState(background: true);
```

Comment:
Background Sync nur im W-Lan. Datenvolumen sparen (auch wenn es nicht viel ist)

Reply:
Dem würde ich stark widersprechen. Viele benutzen die App nur in der Gruppenstunde oder in Leiterrunden. Wir haben dort zum Beispiel kein WLAN, was dazu führt, dass nie ein background sync stattfindet. Und wie du schon meintest, entsprechen die Datenmengen bei einem force sync nicht mal einem Bild oder so. Ein normaler sync ist im Normalfall ja nochmal deutlich kleiner.

Reply:
Sync hast du doch trotzdem. Alle 30 Tage ist er Pflicht. Vor bearbeiten von Daten (Stufenwechsel) sollte auch einer Pflicht werden

Kann man einen Sync mit gedrosseltem Datenvolumen Simulieren? W-Lan normalerweise auch stabiler. Wenn ich jetzt mit der DB quer durch Deutschland unterwegs bin... Würde mich interessieren wie es dann mit Timeouts funktioniert.

Reply:
Man kann die Verbindungsqualität auf jeden Fall im Android Emulator einstellen. Aber das habe ich nicht aufgesetzt. Ich würde die genau wie die `ClientExeption` behandeln. Also sobald eine Anfrage nicht klappt, wird der gesamte Sync abgebrochen und wie wenn man komplett offline ist behandelt.

Ein kompletter Sync vor Stufenwechsel ist eine gute Idee.
Ich sehe immer noch keinen guten Grund, es nicht über mobile Daten zu machen. Wenn man in den 30 Tagen die App immer online über mobile Daten verwendet und dann am Ende einmal wirklich offline verwendet, kann man es nicht, weil es die letzten Male nicht zurückgesetzt wurde.

Reply:
Was hält du von einer Option in den Einstellungen: Daten im Hintergrund über Mobilfunk synchronisieren

Reply:
Finde ich gut als Kompromiss. Ich nehme an, du bist für aus als Standardwert?